src/ball_tree.jl

```diff
@@ -132,22 +129,79 @@ function build_BallTree(index::Int,
 
     # Sort the data at the mid_idx boundary using the split_dim
     # to compare
-    select_spec!(indices, mid_idx, low, high, data, split_dim)
+    select_spec!(indices, mid_idx, low, high, data, split_dim) # culprit? technically, low and high should be disjoint for different threads
 
     build_BallTree(getleft(index), data, data_reordered, hyper_spheres, metric,
-                   indices, indices_reordered, low, mid_idx - 1,
-                   tree_data, array_buffs, reorder)
+                 indices, indices_reordered, low, mid_idx - 1,
+                 tree_data, reorder, parallel)
 
     build_BallTree(getright(index), data, data_reordered, hyper_spheres, metric,
-                  indices, indices_reordered, mid_idx, high,
-                  tree_data, array_buffs, reorder)
+                indices, indices_reordered, mid_idx, high,
+                tree_data, reorder, parallel)
+
+    # Finally create bounding hyper sphere from the two children's hyper spheres
+    hyper_spheres[index]  =  create_bsphere(metric, hyper_spheres[getleft(index)],
+                                            hyper_spheres[getright(index)])
+    return
+end
+
+# Parallelized recursive function to build the tree.
+function build_BallTree(index::Int,
+                        data::Vector{V},
+                        data_reordered::Vector{V},
+                        hyper_spheres::Vector{HyperSphere{N,T}},
+                        metric::Metric,
+                        indices::Vector{Int},
+                        indices_reordered::Vector{Int},
+                        low::Int,
+                        high::Int,
+                        tree_data::TreeData,
+                        reorder::Bool,
+                        parallel::Val{true},
```

Comment:
Using a `Val` and use a separate function like this feels a bit awkward. Couldn't one just look at `parallel_size` in the original `build_BallTree` function  and then decide whether to call the parallel function or the serial one?

Reply:
Using type dispatch on the parallel variable is important, because the compiler is able to get rid of temporary allocations during sequential execution. I can isolate the recursive component of the function though, and only use the `Val(true)` dispatch for that. If we only use a regular `if` statement on a `Bool`, performance during sequential execution will take a hit compared to the status quo.